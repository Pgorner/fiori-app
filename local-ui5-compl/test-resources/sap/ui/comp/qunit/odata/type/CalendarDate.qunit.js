/*global QUnit*/
(function () {
	"use strict";
	QUnit.config.autostart = false;

	sap.ui.define([
		'sap/ui/comp/odata/type/CalendarDate',
		"sap/ui/model/ValidateException"
	], function (CalendarDateType, ValidateException) {

		function fnFormattingParsingHandler(assert, sCalendarType, sInputData) {
			var oCalendarDate = new CalendarDateType(null, null, {
				calendarType: sCalendarType
			});
			var vFormattedValue = oCalendarDate.formatValue(sInputData, "string");
			var vParsedValue = oCalendarDate.parseValue(vFormattedValue, "string");

			assert.equal(sInputData, vParsedValue, "The parsing is reversing the input of the formatting.");
			oCalendarDate.destroy();
			oCalendarDate = null;
		}

		function fnFormattingParsingHandlerWithDigitSequence(assert, sCalendarType, sInputData) {
			var oCalendarDate = new CalendarDateType(null, { isDigitSequence:true }, {
				calendarType: sCalendarType
			});
			var vFormattedValue = oCalendarDate.formatValue(sInputData, "string");
			var vParsedValue = oCalendarDate.parseValue(vFormattedValue, "string");

			if (sInputData !== "000") {
				assert.equal(sInputData, vParsedValue, "The parsing is reversing the input of the formatting.");
			} else {
				assert.equal(null, vParsedValue, "The parsing should return null");
			}
			oCalendarDate.destroy();
			oCalendarDate = null;
		}

		function fnValidationHandler(assert, sCalendarType, sInputData) {
			var oCalendarDate = new CalendarDateType(null, null, {
				calendarType: sCalendarType
			});

			assert.throws(function() {
				oCalendarDate.validateValue(oCalendarDate.parseValue(sInputData, "string"));
			}, function(oError) {
				return oError instanceof ValidateException;
			});
		}

		QUnit.module("CalendarDate Formatting and Parsing");

		QUnit.test("Formatting/Parsing IsCalendarYear", function (assert) {
			fnFormattingParsingHandler(assert, "IsCalendarYear", "2012");
			fnFormattingParsingHandlerWithDigitSequence(assert, "IsCalendarYear", "2012");
		});

		QUnit.test("Formatting/Parsing IsCalendarWeek", function (assert) {
			fnFormattingParsingHandler(assert, "IsCalendarWeek", "51");
			fnFormattingParsingHandlerWithDigitSequence(assert, "IsCalendarWeek", "51");
		});

		QUnit.test("Formatting/Parsing IsCalendarMonth", function (assert) {
			fnFormattingParsingHandler(assert, "IsCalendarMonth", "12");
			fnFormattingParsingHandlerWithDigitSequence(assert, "IsCalendarMonth", "12");
		});


		QUnit.test("Formatting/Parsing IsCalendarYearWeek", function (assert) {
			fnFormattingParsingHandler(assert, "IsCalendarYearWeek", "201251");
			fnFormattingParsingHandlerWithDigitSequence(assert, "IsCalendarYearWeek", "201251");
		});

		QUnit.test("Formatting/Parsing IsCalendarYearMonth", function (assert) {
			fnFormattingParsingHandler(assert, "IsCalendarYearMonth", "202211");
			fnFormattingParsingHandlerWithDigitSequence(assert, "IsCalendarYearMonth", "202211");
		});

		QUnit.test("Formatting/Parsing IsCalendarYearQuarter", function (assert) {
			fnFormattingParsingHandler(assert, "IsCalendarYearQuarter", "20123");
			fnFormattingParsingHandlerWithDigitSequence(assert, "IsCalendarYearQuarter", "20123");
		});


		QUnit.module("CalendarDate Validation");

		QUnit.test("Formatting/Parsing IsCalendarYear", function (assert) {
			fnValidationHandler(assert, "IsCalendarYear", "-2022");
			fnValidationHandler(assert, "IsCalendarYear", "0123");
			fnValidationHandler(assert, "IsCalendarYear", "12345");
		});

		QUnit.test("Formatting/Parsing IsCalendarWeek", function (assert) {
			fnValidationHandler(assert, "IsCalendarWeek", "-1");
			fnValidationHandler(assert, "IsCalendarWeek", "54");
			fnValidationHandler(assert, "IsCalendarWeek", "0");
		});

		QUnit.test("Formatting/Parsing IsCalendarMonth", function (assert) {
			fnValidationHandler(assert, "IsCalendarMonth", "-1");
			fnValidationHandler(assert, "IsCalendarMonth", "13");
			fnValidationHandler(assert, "IsCalendarMonth", "0");
		});

		QUnit.test("Formatting/Parsing IsCalendarQuarter", function (assert) {
			fnValidationHandler(assert, "IsCalendarQuarter", "-1");
			fnValidationHandler(assert, "IsCalendarQuarter", "11");
			fnValidationHandler(assert, "IsCalendarQuarter", "5");
			fnValidationHandler(assert, "IsCalendarQuarter", "0");
		});

		QUnit.test("Formatting/Parsing IsCalendarYearWeek", function (assert) {
			fnValidationHandler(assert, "IsCalendarYearWeek", "-202253");
			fnValidationHandler(assert, "IsCalendarYearWeek", "001224");
			fnValidationHandler(assert, "IsCalendarYearWeek", "2022324");
			fnValidationHandler(assert, "IsCalendarYearWeek", "202254");
			fnValidationHandler(assert, "IsCalendarYearWeek", "20220");
			fnValidationHandler(assert, "IsCalendarYearWeek", "2022544");
		});

		QUnit.test("Formatting/Parsing IsCalendarYearMonth", function (assert) {
			fnValidationHandler(assert, "IsCalendarYearMonth", "-202212");
			fnValidationHandler(assert, "IsCalendarYearMonth", "001212");
			fnValidationHandler(assert, "IsCalendarYearMonth", "2022314");
			fnValidationHandler(assert, "IsCalendarYearMonth", "202213");
			fnValidationHandler(assert, "IsCalendarYearMonth", "20220");
			fnValidationHandler(assert, "IsCalendarYearMonth", "2022144");
		});

		QUnit.test("Formatting/Parsing IsCalendarYearQuarter", function (assert) {
			fnValidationHandler(assert, "IsCalendarYearQuarter", "-20224");
			fnValidationHandler(assert, "IsCalendarYearQuarter", "00124");
			fnValidationHandler(assert, "IsCalendarYearQuarter", "202254");
			fnValidationHandler(assert, "IsCalendarYearQuarter", "202221");
			fnValidationHandler(assert, "IsCalendarYearQuarter", "20225");
			fnValidationHandler(assert, "IsCalendarYearQuarter", "20220");
		});

		QUnit.test("CalendarDate#getFormatter returns an instance of PeriodDateFormat", function(assert) {
			var oFormatter, oCalendarDate;

			oCalendarDate = new CalendarDateType(null, null, {
				calendarType: "IsCalendarYear"
			});
			assert.ok(typeof oCalendarDate.getFormatter === "function", "Function getFormatter exists");

			oFormatter = oCalendarDate.getFormatter();
			assert.ok(typeof oFormatter === "object", "Formatter object returned");
			assert.ok(typeof oFormatter.isA === "function", "Formatter extends sap/ui/base/Object");
			assert.ok(oFormatter.isA("sap.ui.comp.odata.PeriodDateFormat"), "Formatter is an instance of sap/ui/comp/odata/CalendarFormat");
		});

		QUnit.start();
	});

})();
