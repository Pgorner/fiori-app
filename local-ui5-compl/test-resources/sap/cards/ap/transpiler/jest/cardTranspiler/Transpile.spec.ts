/*!
 * SAP UI development toolkit for HTML5 (SAPUI5)
 *  * (c) Copyright 2009-2025 SAP SE. All rights reserved
 */
import { convertIntegrationCardToAdaptive } from "sap/cards/ap/transpiler/cardTranspiler/Transpile";
import type { CardManifest } from "sap/ui/integration/widgets/Card";
import integrationObjectCard from "../../cards/integration/IntegrationObjectCard";

describe("Transpile: generateAdaptiveCard", () => {
	let windowSpy: jest.SpyInstance;

	beforeEach(() => {
		windowSpy = jest.spyOn(window, "window", "get");
		windowSpy.mockImplementation(() => ({
			hasher: {
				getHash: () => "SalesOrder-manageV2?SalesOrderDate=2024-03-01&/SalesOrderManage('400')"
			},
			location: {
				href: "https://my313815.s4hana.ondemand.com/ui#SalesOrder-manageV2?SalesOrderDate=2024-03-01&/SalesOrderManage('400')",
				origin: "https://my313815.s4hana.ondemand.com"
			}
		}));
	});

	afterEach(() => {
		windowSpy.mockRestore();
		jest.clearAllMocks();
	});

	test("Call with empty integration card", () => {
		const integrationCardManifest = {
			"sap.card": {
				configuration: {
					parameters: {}
				}
			}
		};
		expect(convertIntegrationCardToAdaptive(integrationCardManifest, "SalesOrder-manageV2")).toMatchSnapshot();
	});
	test("Generate adaptive Object card", () => {
		// Integration card is transpiled into adaptive card
		expect(
			convertIntegrationCardToAdaptive(integrationObjectCard.manifest as unknown as CardManifest, "SalesOrder-manageV2")
		).toMatchSnapshot();
	});

	test("Generate adaptive Object card with formatter Expressions in manifest value", () => {
		// Integration card is transpiled into adaptive card
		expect(
			convertIntegrationCardToAdaptive(
				integrationObjectCard.manifestWithExpressions as unknown as CardManifest,
				"SalesOrder-manageV2"
			)
		).toMatchSnapshot();
	});

	test("Generate adaptive Object card when manifest may or may not have values for group items", () => {
		// Integration card is transpiled into adaptive card correctly
		const keyParameters = [
			{
				key: "SalesOrderDate",
				formattedValue: "2024-03-01"
			},
			{
				key: "SalesOrder",
				formattedValue: "1234"
			}
		];
		expect(
			convertIntegrationCardToAdaptive(
				integrationObjectCard.manifestWithOrWithoutGroupItemValues as unknown as CardManifest,
				"SalesOrder-manageV2",
				keyParameters
			)
		).toMatchSnapshot();
	});
});
