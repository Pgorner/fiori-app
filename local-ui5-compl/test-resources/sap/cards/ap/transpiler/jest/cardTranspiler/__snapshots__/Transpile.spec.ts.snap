// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transpile: generateAdaptiveCard Call with empty integration card 1`] = `
{
  "$schema": "https://adaptivecards.io/schemas/adaptive-card.json",
  "body": [],
  "metadata": {},
  "type": "AdaptiveCard",
  "version": "1.6",
}
`;

exports[`Transpile: generateAdaptiveCard Generate adaptive Object card 1`] = `
{
  "$schema": "https://adaptivecards.io/schemas/adaptive-card.json",
  "body": [
    {
      "color": "Default",
      "maxLines": 1,
      "size": "default",
      "spacing": "none",
      "text": "[ \${{salesOrderHeaderTitle}} ](https://my313815.s4hana.ondemand.com/ui#SalesOrder-manageV2&/C_STTA_SalesOrder_WD_20(node_key='fa163ee4-7bdd-1ee8-b1ff-d3c5a4e5236a',IsActiveEntity='true'))",
      "type": "TextBlock",
      "weight": "bolder",
      "wrap": false,
    },
    {
      "columns": [
        {
          "items": [
            {
              "$when": "\${ ('{so_id}') ? true: false}",
              "color": "Default",
              "isSubtle": true,
              "maxLines": 1,
              "size": "default",
              "spacing": "none",
              "text": "\${so_id}",
              "type": "TextBlock",
              "weight": "default",
              "wrap": false,
            },
          ],
          "type": "Column",
          "width": "auto",
        },
        {
          "items": [
            {
              "$when": "\${ ('{so_id}' && '{currency_code}') ? true: false}",
              "color": "Default",
              "horizontalAlignment": "Center",
              "isSubtle": true,
              "maxLines": 1,
              "size": "default",
              "spacing": "none",
              "text": "|",
              "type": "TextBlock",
              "weight": "default",
              "wrap": false,
            },
          ],
          "type": "Column",
          "width": "auto",
        },
        {
          "items": [
            {
              "$when": "\${ ('{currency_code}') ? true: false}",
              "color": "Default",
              "isSubtle": true,
              "maxLines": 1,
              "size": "default",
              "spacing": "none",
              "text": "\${currency_code}",
              "type": "TextBlock",
              "weight": "default",
              "wrap": false,
            },
          ],
          "type": "Column",
          "width": "auto",
        },
      ],
      "spacing": "none",
      "type": "ColumnSet",
    },
    {
      "color": "Default",
      "maxLines": 1,
      "size": "extraLarge",
      "spacing": "none",
      "text": "\${overall_status}",
      "type": "TextBlock",
      "weight": "default",
      "wrap": false,
    },
    {
      "items": [
        {
          "items": [
            {
              "color": "Default",
              "maxLines": 2,
              "size": "small",
              "spacing": "medium",
              "text": "General Information",
              "type": "TextBlock",
              "weight": "bolder",
              "wrap": true,
            },
            {
              "items": [
                {
                  "columns": [
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "{{netAmountLabel}}:",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "\${if(trim(string(net_amount)) == '', '&minus;', string(net_amount))}",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "Tax Amount:",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "\${if(trim(string(tax_amount)) == '', '&minus;', string(tax_amount))}",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "Gross Amount:",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "\${if(trim(string(gross_amount)) == '', '&minus;', string(gross_amount))}",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                  ],
                  "type": "ColumnSet",
                },
              ],
              "spacing": "small",
              "type": "Container",
            },
          ],
          "spacing": "medium",
          "type": "Container",
        },
        {
          "items": [
            {
              "color": "Default",
              "maxLines": 2,
              "size": "small",
              "spacing": "medium",
              "text": "Additional Information",
              "type": "TextBlock",
              "weight": "bolder",
              "wrap": true,
            },
            {
              "items": [
                {
                  "columns": [
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "{{productLabel}}:",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "\${if(trim(string(Product)) == '', '&minus;', string(Product))}",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "Language:",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "\${if(trim(string(Language)) == '', '&minus;', string(Language))}",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "type": "Column",
                      "width": "stretch",
                    },
                  ],
                  "type": "ColumnSet",
                },
              ],
              "spacing": "small",
              "type": "Container",
            },
          ],
          "spacing": "medium",
          "type": "Container",
        },
      ],
      "spacing": "large",
      "type": "Container",
    },
    {
      "actions": [
        {
          "$when": "\${__OperationControl.SetBillingBlock ? true : false}",
          "card": {
            "actions": [
              {
                "data": {
                  "contextURL": "https://my313815.s4hana.ondemand.com/sap/opu/odata/sap/salesorder/C_STTA_SalesOrder_WD_20(node_key='fa163ee4-7bdd-1ee8-b1ff-d3c5a4e5236a',IsActiveEntity='true')",
                  "isConfirmationRequired": false,
                },
                "style": "positive",
                "title": "OK",
                "type": "Action.Execute",
                "verb": "com.sap.gateway.srvd.ui_creditmemorequestmanage.v0001.SetBillingBlock",
              },
            ],
            "body": [
              {
                "choices": [
                  {
                    "$data": "\${I_BillingBlockReason}",
                    "title": "\${BillingBlockReason_Text} (\${BillingBlockReason})",
                    "value": "\${BillingBlockReason}",
                  },
                ],
                "errorMessage": "This is a required input",
                "id": "HeaderBillingBlockReason",
                "isMultiline": false,
                "isRequired": true,
                "label": "HeaderBillingBlockReason",
                "placeholder": "Please choose",
                "type": "Input.ChoiceSet",
              },
            ],
            "type": "AdaptiveCard",
          },
          "isEnabled": "\${__OperationControl.SetBillingBlock}",
          "style": "default",
          "title": "Set Billing Block",
          "type": "Action.ShowCard",
        },
        {
          "$when": "\${salesorderID ? true : false}",
          "data": {
            "actionParams": {
              "keys": [
                "SalesPlanUUID",
                "IsActiveEntity",
              ],
            },
            "contextURL": "https://my313815.s4hana.ondemand.com/sap/opu/odata/sap/salesorder/C_STTA_SalesOrder_WD_20(node_key='fa163ee4-7bdd-1ee8-b1ff-d3c5a4e5236a',IsActiveEntity='true')",
            "isConfirmationRequired": false,
          },
          "isEnabled": "\${salesorderID}",
          "style": "default",
          "title": "Release",
          "type": "Action.Execute",
          "verb": "C_SalesPlanTPRelease",
        },
      ],
      "type": "ActionSet",
    },
  ],
  "metadata": {
    "sap.cards.ap": {
      "cardType": "LEAN_DT",
      "parentAppId": "cus.sd.salesorderv2.manage",
      "templateName": "ObjectPage",
      "versions": {
        "ui5": "1.120.1-202405021611",
      },
    },
    "webUrl": "https://my313815.s4hana.ondemand.com/sap/opu/odata/sap/salesorder/C_STTA_SalesOrder_WD_20(node_key='fa163ee4-7bdd-1ee8-b1ff-d3c5a4e5236a',IsActiveEntity='true')",
  },
  "msteams": {
    "width": "full",
  },
  "type": "AdaptiveCard",
  "version": "1.6",
}
`;

exports[`Transpile: generateAdaptiveCard Generate adaptive Object card when manifest may or may not have values for group items 1`] = `
{
  "$schema": "https://adaptivecards.io/schemas/adaptive-card.json",
  "body": [
    {
      "color": "Default",
      "maxLines": 1,
      "size": "default",
      "spacing": "none",
      "text": "",
      "type": "TextBlock",
      "weight": "bolder",
      "wrap": false,
    },
    {
      "columns": [
        {
          "items": [
            {
              "$when": "\${ (false) ? true: false}",
              "color": "Default",
              "isSubtle": true,
              "maxLines": 1,
              "size": "default",
              "spacing": "none",
              "text": "",
              "type": "TextBlock",
              "weight": "default",
              "wrap": false,
            },
          ],
          "type": "Column",
          "width": "auto",
        },
        {
          "items": [
            {
              "$when": "\${ ( false) ? true: false}",
              "color": "Default",
              "horizontalAlignment": "Center",
              "isSubtle": true,
              "maxLines": 1,
              "size": "default",
              "spacing": "none",
              "text": "|",
              "type": "TextBlock",
              "weight": "default",
              "wrap": false,
            },
          ],
          "type": "Column",
          "width": "auto",
        },
        {
          "items": [
            {
              "$when": "\${ (false) ? true: false}",
              "color": "Default",
              "isSubtle": true,
              "maxLines": 1,
              "size": "default",
              "spacing": "none",
              "text": "",
              "type": "TextBlock",
              "weight": "default",
              "wrap": false,
            },
          ],
          "type": "Column",
          "width": "auto",
        },
      ],
      "spacing": "none",
      "type": "ColumnSet",
    },
    {
      "color": "Default",
      "maxLines": 1,
      "size": "extraLarge",
      "spacing": "none",
      "text": "",
      "type": "TextBlock",
      "weight": "default",
      "wrap": false,
    },
    {
      "items": [
        {
          "items": [
            {
              "color": "Default",
              "maxLines": 2,
              "size": "small",
              "spacing": "medium",
              "text": "Group 1",
              "type": "TextBlock",
              "weight": "bolder",
              "wrap": true,
            },
            {
              "items": [
                {
                  "columns": [
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "Gross Amount:",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "\${gross_amount} \${currency_code}",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "type": "Column",
                      "width": "stretch",
                    },
                  ],
                  "type": "ColumnSet",
                },
              ],
              "spacing": "small",
              "type": "Container",
            },
          ],
          "spacing": "medium",
          "type": "Container",
        },
        {
          "items": [
            {
              "color": "Default",
              "maxLines": 2,
              "size": "small",
              "spacing": "medium",
              "text": "Group 2",
              "type": "TextBlock",
              "weight": "bolder",
              "wrap": true,
            },
            {
              "items": [
                {
                  "columns": [
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "Net Amount:",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "\${net_amount} \${currency_code}",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "type": "Column",
                      "width": "stretch",
                    },
                  ],
                  "type": "ColumnSet",
                },
              ],
              "spacing": "small",
              "type": "Container",
            },
          ],
          "spacing": "medium",
          "type": "Container",
        },
        {
          "items": [
            {
              "color": "Default",
              "maxLines": 2,
              "size": "small",
              "spacing": "medium",
              "text": "Group 3",
              "type": "TextBlock",
              "weight": "bolder",
              "wrap": true,
            },
            {
              "items": [
                {
                  "columns": [
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "type": "Column",
                      "width": "stretch",
                    },
                  ],
                  "type": "ColumnSet",
                },
              ],
              "spacing": "small",
              "type": "Container",
            },
          ],
          "spacing": "medium",
          "type": "Container",
        },
      ],
      "spacing": "large",
      "type": "Container",
    },
  ],
  "metadata": {
    "sap.cards.ap": undefined,
    "webUrl": undefined,
  },
  "msteams": {
    "width": "full",
  },
  "type": "AdaptiveCard",
  "version": "1.6",
}
`;

exports[`Transpile: generateAdaptiveCard Generate adaptive Object card with formatter Expressions in manifest value 1`] = `
{
  "$schema": "https://adaptivecards.io/schemas/adaptive-card.json",
  "body": [
    {
      "color": "Default",
      "maxLines": 1,
      "size": "default",
      "spacing": "none",
      "text": "[ \${{salesOrderHeaderTitle}} ](https://my313815.s4hana.ondemand.com/ui#SalesOrder-manageV2&/C_STTA_SalesOrder_WD_20(node_key='fa163ee4-7bdd-1ee8-b1ff-d3c5a4e5236a',IsActiveEntity='true'))",
      "type": "TextBlock",
      "weight": "bolder",
      "wrap": false,
    },
    {
      "columns": [
        {
          "items": [
            {
              "$when": "\${ ('{so_id}') ? true: false}",
              "color": "Default",
              "isSubtle": true,
              "maxLines": 1,
              "size": "default",
              "spacing": "none",
              "text": "\${so_id}",
              "type": "TextBlock",
              "weight": "default",
              "wrap": false,
            },
          ],
          "type": "Column",
          "width": "auto",
        },
        {
          "items": [
            {
              "$when": "\${ ('{so_id}' && '{currency_code}') ? true: false}",
              "color": "Default",
              "horizontalAlignment": "Center",
              "isSubtle": true,
              "maxLines": 1,
              "size": "default",
              "spacing": "none",
              "text": "|",
              "type": "TextBlock",
              "weight": "default",
              "wrap": false,
            },
          ],
          "type": "Column",
          "width": "auto",
        },
        {
          "items": [
            {
              "$when": "\${ ('{currency_code}') ? true: false}",
              "color": "Default",
              "isSubtle": true,
              "maxLines": 1,
              "size": "default",
              "spacing": "none",
              "text": "\${currency_code}",
              "type": "TextBlock",
              "weight": "default",
              "wrap": false,
            },
          ],
          "type": "Column",
          "width": "auto",
        },
      ],
      "spacing": "none",
      "type": "ColumnSet",
    },
    {
      "color": "\${if(toLower(net_amount) == 'critical', 'Warning', if(toLower(net_amount) == 'error', 'Attention', if(toLower(net_amount) == 'good', 'Good', if(int(net_amount) == 3, 'Good', if(int(net_amount) == 1, 'Attention', if(int(net_amount) == 2, 'Warning', 'Default'))))))}",
      "maxLines": 1,
      "size": "extraLarge",
      "spacing": "none",
      "text": "\${gross_amount}",
      "type": "TextBlock",
      "weight": "default",
      "wrap": false,
    },
    {
      "items": [
        {
          "items": [
            {
              "color": "Default",
              "maxLines": 2,
              "size": "small",
              "spacing": "medium",
              "text": "General Information",
              "type": "TextBlock",
              "weight": "bolder",
              "wrap": true,
            },
            {
              "items": [
                {
                  "columns": [
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "{{netAmountLabel}}:",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "\${if(trim(string(net_amount)) == '', '&minus;', string(net_amount))}",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "items": [
                        {
                          "color": "Default",
                          "isSubtle": true,
                          "maxLines": 1,
                          "size": "small",
                          "spacing": "default",
                          "text": "Created on:",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": false,
                        },
                        {
                          "color": "Default",
                          "maxLines": 2,
                          "size": "small",
                          "spacing": "none",
                          "text": "\${if(CreationDateTime && CreationDateTime != '', convertToUTC(CreationDateTime, 'UTC', 'ddd d MMM YYYY HH:mm'), '&minus;')}",
                          "type": "TextBlock",
                          "weight": "default",
                          "wrap": true,
                        },
                      ],
                      "type": "Column",
                      "width": "stretch",
                    },
                    {
                      "type": "Column",
                      "width": "stretch",
                    },
                  ],
                  "type": "ColumnSet",
                },
              ],
              "spacing": "small",
              "type": "Container",
            },
          ],
          "spacing": "medium",
          "type": "Container",
        },
      ],
      "spacing": "large",
      "type": "Container",
    },
  ],
  "metadata": {
    "sap.cards.ap": {
      "cardType": "LEAN_DT",
      "parentAppId": "sales.order.wd20",
      "templateName": "ObjectPage",
      "versions": {
        "ui5": "1.120.1-202405021611",
      },
    },
    "webUrl": "https://my313815.s4hana.ondemand.com/sap/opu/odata/sap/salesorder/C_STTA_SalesOrder_WD_20(node_key='fa163ee4-7bdd-1ee8-b1ff-d3c5a4e5236a',IsActiveEntity='true')",
  },
  "msteams": {
    "width": "full",
  },
  "type": "AdaptiveCard",
  "version": "1.6",
}
`;
